File: frontend-web/src/types/currencyExchange.ts
// Currency Exchange System Types

// Enums
export enum CurrencyType {
  HTG = 1,
  USD = 2,
  EUR = 3,
  CAD = 4,
  DOP = 5,
  JMD = 6
}

export enum ExchangeType {
  Purchase = 1,  // Client achète des devises (Banque vend)
  Sale = 2       // Client vend des devises (Banque achète)
}

export enum RateUpdateMethod {
  Manual = 1,
  Automatic = 2,
  External = 3
}

export enum ExchangeTransactionStatus {
  Pending = 1,
  Completed = 2,
  Cancelled = 3,
  Failed = 4
}

export enum CurrencyMovementType {
  Purchase = 0,   // Acha
  Sale = 1,       // Vant
  Transfer = 2,   // Transfè
  Reserve = 3,    // Rezèv
  Adjustment = 4  // Ajisteman
}

// Base Interfaces
export interface CurrencyExchangeRate {
  id: string;
  baseCurrency: CurrencyType;
  baseCurrencyName: string;
  targetCurrency: CurrencyType;
  targetCurrencyName: string;
  buyingRate: number;
  sellingRate: number;
  effectiveDate: string;
  expiryDate?: string;
  updateMethod: RateUpdateMethod;
  updateMethodName: string;
  isActive: boolean;
  notes?: string;
  createdBy: string;
  createdByName: string;
  updatedBy?: string;
  updatedByName?: string;
  createdAt: string;
  updatedAt: string;
  transactions?: ExchangeTransaction[];
}

export interface ExchangeTransaction {
  id: string;
  transactionNumber: string;
  branchId: string;
  branchName: string;
  exchangeRateId: string;
  exchangeType: ExchangeType;
  exchangeTypeName: string;
  fromCurrency: CurrencyType;
  fromCurrencyName: string;
  toCurrency: CurrencyType;
  toCurrencyName: string;
  fromAmount: number;
  toAmount: number;
  exchangeRate: number;
  commissionAmount: number;
  commissionRate: number;
  netAmount: number;
  customerName?: string;
  customerDocument?: string;
  customerPhone?: string;
  status: ExchangeTransactionStatus;
  statusName: string;
  transactionDate: string;
  processedBy: string;
  processedByName: string;
  approvedBy?: string;
  notes?: string;
  receiptNumber?: string;
  receiptPrinted: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CurrencyReserve {
  id: string;
  branchId: string;
  branchName: string;
  currency: CurrencyType;
  currencyName: string;
  currentBalance: number;
  minimumThreshold: number;
  maximumThreshold: number;
  reorderLevel: number;
  lastUpdated: string;
  updatedBy: string;
  updatedByName: string;
  movements?: CurrencyMovement[];
}

export interface CurrencyMovement {
  id: string;
  branchId: string;
  branchName: string;
  currency: CurrencyType;
  currencyName: string;
  movementType: CurrencyMovementType;
  movementTypeName: string;
  amount: number;
  balanceBefore: number;
  balanceAfter: number;
  exchangeTransactionId?: string;
  reference?: string;
  description: string;
  processedBy: string;
  processedByName: string;
  approvedBy?: string;
  movementDate: string;
  createdAt: string;
}

// DTOs for forms and API calls
export interface CreateExchangeRateDto {
  baseCurrency: CurrencyType;
  targetCurrency: CurrencyType;
  buyingRate: number;
  sellingRate: number;
  effectiveDate: string;
  expiryDate?: string;
  updateMethod: RateUpdateMethod;
  notes?: string;
}

export interface UpdateExchangeRateDto extends Partial<CreateExchangeRateDto> {
  id: string;
  isActive?: boolean;
}

export interface ExchangeCalculationDto {
  branchId: string;
  exchangeType: ExchangeType;
  amount: number;
}

export interface ExchangeCalculationResult {
  fromCurrency: CurrencyType;
  fromCurrencyName: string;
  toCurrency: CurrencyType;
  toCurrencyName: string;
  fromAmount: number;
  toAmount: number;
  exchangeRate: number;
  commissionRate: number;
  commissionAmount: number;
  netAmount: number;
  rateId: string;
  isValid: boolean;
  message?: string;
  errorMessage?: string;
  availableBalance?: number;
}

export interface ProcessExchangeDto {
  branchId: string;
  fromCurrency: CurrencyType;
  toCurrency: CurrencyType;
  fromAmount: number;
  exchangeType: ExchangeType;
  customerName?: string;
  customerDocument?: string;
  customerPhone?: string;
  notes?: string;
}

export interface UpdateCurrencyReserveDto {
  id: string;
  minimumThreshold?: number;
  maximumThreshold?: number;
  reorderLevel?: number;
}

export interface CreateCurrencyMovementDto {
  branchId: string;
  currency: CurrencyType;
  movementType: CurrencyMovementType;
  amount: number;
  reference?: string;
  description: string;
}

// Search and Filter DTOs
export interface ExchangeRateSearchDto {
  baseCurrency?: CurrencyType;
  targetCurrency?: CurrencyType;
  isActive?: boolean;
  effectiveDate?: string;
  page?: number;
  pageSize?: number;
}

export interface ExchangeTransactionSearchDto {
  branchId?: string;
  exchangeType?: ExchangeType;
  fromCurrency?: CurrencyType;
  toCurrency?: CurrencyType;
  status?: ExchangeTransactionStatus;
  startDate?: string;
  endDate?: string;
  includeAll?: boolean;
  customerName?: string;
  transactionNumber?: string;
  page?: number;
  pageSize?: number;
}

export interface CurrencyMovementSearchDto {
  branchId?: string;
  currency?: CurrencyType;
  movementType?: CurrencyMovementType;
  startDate?: string;
  endDate?: string;
  page?: number;
  pageSize?: number;
}

// Statistics and Reports
export interface ExchangeStatistics {
  totalTransactions: number;
  totalVolume: number;
  totalCommission: number;
  averageTransactionAmount: number;
  transactionsByType: {
    exchangeType: ExchangeType;
    exchangeTypeName: string;
    count: number;
    volume: number;
    commission: number;
  }[];
  transactionsByCurrency: {
    currency: CurrencyType;
    currencyName: string;
    buyCount: number;
    sellCount: number;
    buyVolume: number;
    sellVolume: number;
    totalCommission: number;
  }[];
  transactionsByBranch: {
    branchId: string;
    branchName: string;
    transactionCount: number;
    totalVolume: number;
    totalCommission: number;
  }[];
}

export interface CurrencyReserveReport {
  branchId: string;
  branchName: string;
  reserves: {
    currency: CurrencyType;
    currencyName: string;
    currentBalance: number;
    minimumThreshold: number;
    maximumThreshold: number;
    availableCapacity: number;
    utilizationPercentage: number;
    needsReorder: boolean;
    status: 'Normal' | 'Low' | 'Critical' | 'Excess';
  }[];
  totalReserveValue: number; // en HTG
  lowReserveCount: number;
  criticalReserveCount: number;
}

export interface ProfitabilityReport {
  period: string;
  totalRevenue: number;
  totalCosts: number;
  netProfit: number;
  profitMargin: number;
  revenueByBranch: {
    branchId: string;
    branchName: string;
    revenue: number;
    transactionCount: number;
    averageMargin: number;
  }[];
  revenueByCurrency: {
    currency: CurrencyType;
    currencyName: string;
    revenue: number;
    volume: number;
    averageSpread: number;
  }[];
}

// Helper functions for formatting
export const formatCurrencyType = (currency: CurrencyType): string => {
  // Some legacy payloads may send numeric 0 for HTG; normalize defensively
  const normalized: CurrencyType = (currency as number) === 0 ? CurrencyType.HTG : currency;
  const currencies = {
    [CurrencyType.HTG]: 'Gourde Haïtienne (HTG)',
    [CurrencyType.USD]: 'Dollar Américain (USD)',
    [CurrencyType.EUR]: 'Euro (EUR)',
    [CurrencyType.CAD]: 'Dollar Canadien (CAD)',
    [CurrencyType.DOP]: 'Peso Dominicain (DOP)',
    [CurrencyType.JMD]: 'Dollar Jamaïcain (JMD)'
  };
  return currencies[normalized] || 'Inconnu';
};

export const formatCurrencySymbol = (currency: CurrencyType): string => {
  const normalized: CurrencyType = (currency as number) === 0 ? CurrencyType.HTG : currency;
  const symbols = {
    [CurrencyType.HTG]: 'HTG',
    [CurrencyType.USD]: '$',
    [CurrencyType.EUR]: '€',
    [CurrencyType.CAD]: 'CAD$',
    [CurrencyType.DOP]: 'RD$',
    [CurrencyType.JMD]: 'J$'
  };
  return symbols[normalized] || '';
};

export const formatExchangeType = (type: ExchangeType): string => {
  const types = {
    [ExchangeType.Purchase]: 'Achat (client achète)',
    [ExchangeType.Sale]: 'Vente (client vend)'
  };
  return types[type] || 'Inconnu';
};

export const formatRateUpdateMethod = (method: RateUpdateMethod): string => {
  // Normalize legacy numeric 0 to Manual
  const normalized: RateUpdateMethod = (method as number) === 0 ? RateUpdateMethod.Manual : method;
  const methods = {
    [RateUpdateMethod.Manual]: 'Manuel',
    [RateUpdateMethod.Automatic]: 'Automatique',
    [RateUpdateMethod.External]: 'Externe'
  };
  return methods[normalized] || 'Inconnu';
};

export const formatTransactionStatus = (status: ExchangeTransactionStatus): string => {
  const statuses = {
    [ExchangeTransactionStatus.Pending]: 'En attente',
    [ExchangeTransactionStatus.Completed]: 'Complété',
    [ExchangeTransactionStatus.Cancelled]: 'Annulé',
    [ExchangeTransactionStatus.Failed]: 'Échoué'
  };
  return statuses[status] || 'Inconnu';
};

export const formatMovementType = (type: CurrencyMovementType): string => {
  const types = {
    [CurrencyMovementType.Purchase]: 'Achat',
    [CurrencyMovementType.Sale]: 'Vente',
    [CurrencyMovementType.Transfer]: 'Transfert',
    [CurrencyMovementType.Reserve]: 'Réserve',
    [CurrencyMovementType.Adjustment]: 'Ajustement'
  };
  return types[type] || 'Inconnu';
};

export const formatCurrencyAmount = (amount: number, currency: CurrencyType): string => {
  const symbol = formatCurrencySymbol(currency);
  if (currency === CurrencyType.HTG) {
    return `${amount.toLocaleString('fr-HT')} ${symbol}`;
  }
  return `${symbol}${amount.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
};

export const getReserveStatus = (currentBalance: number, minimumThreshold: number, maximumThreshold: number): string => {
  if (currentBalance <= minimumThreshold * 0.5) return 'Critical';
  if (currentBalance <= minimumThreshold) return 'Low';
  if (currentBalance >= maximumThreshold) return 'Excess';
  return 'Normal';
};

export const getReserveStatusColor = (status: string): string => {
  const colors: Record<string, string> = {
    'Normal': 'green',
    'Low': 'yellow',
    'Critical': 'red',
    'Excess': 'blue'
  };
  return colors[status] || 'gray';
};

File: backend/NalaCreditAPI/Models/CurrencyExchange.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace NalaCreditAPI.Models
{
    public enum CurrencyType
    {
        HTG = 1, // Haitian Gourdes
        USD = 2  // US Dollar
    }

    public enum ExchangeType
    {
        Purchase = 1,  // Achte devises (HTG → USD)
        Sale = 2       // Vann devises (USD → HTG)
    }

    public enum ExchangeTransactionStatus
    {
        Pending = 1,
        Completed = 2,
        Cancelled = 3,
        Failed = 4
    }

    public enum RateUpdateMethod
    {
        Manual = 1,
        Automatic = 2
    }

    public class CurrencyExchangeRate
    {
        public Guid Id { get; set; }

        [Required]
        public CurrencyType BaseCurrency { get; set; } = CurrencyType.HTG;

        [Required]
        public CurrencyType TargetCurrency { get; set; } = CurrencyType.USD;

        [Required]
        [Column(TypeName = "decimal(18,6)")]
        public decimal BuyingRate { get; set; } // Taux d'achat

        [Required]
        [Column(TypeName = "decimal(18,6)")]
        public decimal SellingRate { get; set; } // Taux de vente

        [Required]
        public DateTime EffectiveDate { get; set; }

        public DateTime? ExpiryDate { get; set; }

        [Required]
        public RateUpdateMethod UpdateMethod { get; set; }

        [Required]
        public bool IsActive { get; set; } = true;

        [StringLength(500)]
        public string Notes { get; set; } = string.Empty;

        [Required]
        [StringLength(100)]
        public string CreatedBy { get; set; } = string.Empty;

        [Required]
        [StringLength(100)]
        public string UpdatedBy { get; set; } = string.Empty;

        [Required]
        public DateTime CreatedAt { get; set; }

        [Required]
        public DateTime UpdatedAt { get; set; }

        // Navigation Properties
        public virtual ICollection<ExchangeTransaction> Transactions { get; set; } = new List<ExchangeTransaction>();
    }

    public class ExchangeTransaction
    {
        public Guid Id { get; set; }

        [Required]
        [StringLength(50)]
        public string TransactionNumber { get; set; } = string.Empty;

        [Required]
        public Guid BranchId { get; set; }

        [Required]
        [StringLength(200)]
        public string BranchName { get; set; } = string.Empty;

        [Required]
        public Guid ExchangeRateId { get; set; }

        [Required]
        public ExchangeType ExchangeType { get; set; }

        [Required]
        public CurrencyType FromCurrency { get; set; }

        [Required]
        public CurrencyType ToCurrency { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal FromAmount { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal ToAmount { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,6)")]
        public decimal ExchangeRate { get; set; }

        [Column(TypeName = "decimal(18,2)")]
        public decimal CommissionAmount { get; set; }

        [Column(TypeName = "decimal(5,4)")]
        public decimal CommissionRate { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal NetAmount { get; set; } // Montant reçu par le client

        [StringLength(200)]
        public string CustomerName { get; set; } = string.Empty;

        [StringLength(50)]
        public string CustomerDocument { get; set; } = string.Empty;

        [StringLength(15)]
        public string CustomerPhone { get; set; } = string.Empty;

        [Required]
        public ExchangeTransactionStatus Status { get; set; }

        [Required]
        public DateTime TransactionDate { get; set; }

        [StringLength(100)]
        public string ProcessedBy { get; set; } = string.Empty;

        [StringLength(100)]
        public string ProcessedByName { get; set; } = string.Empty;

        [StringLength(100)]
        public string ApprovedBy { get; set; } = string.Empty;

        [StringLength(500)]
        public string Notes { get; set; } = string.Empty;

        [StringLength(100)]
        public string ReceiptNumber { get; set; } = string.Empty;

        public bool ReceiptPrinted { get; set; } = false;

        [Required]
        public DateTime CreatedAt { get; set; }

        [Required]
        public DateTime UpdatedAt { get; set; }

        // Navigation Properties
        public virtual CurrencyExchangeRate CurrencyRate { get; set; } = null!;
    }

    public class CurrencyReserve
    {
        public Guid Id { get; set; }

        [Required]
        public Guid BranchId { get; set; }

        [Required]
        [StringLength(200)]
        public string BranchName { get; set; } = string.Empty;

        [Required]
        public CurrencyType Currency { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal CurrentBalance { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal MinimumBalance { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal MaximumBalance { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal DailyLimit { get; set; }

        [Column(TypeName = "decimal(18,2)")]
        public decimal DailyUsed { get; set; }

        public DateTime LastRestockDate { get; set; }

        public DateTime LastDepositDate { get; set; }

        [Required]
        public bool IsActive { get; set; } = true;

        [StringLength(500)]
        public string Notes { get; set; } = string.Empty;

        [Required]
        public DateTime CreatedAt { get; set; }

        [Required]
        public DateTime UpdatedAt { get; set; }

        [Required]
        [StringLength(100)]
        public string UpdatedBy { get; set; } = string.Empty;

        // Navigation Properties
        public virtual ICollection<CurrencyMovement> CurrencyMovements { get; set; } = new List<CurrencyMovement>();
    }

    public enum CurrencyMovementType
    {
        Restock = 1,        // Approvisionnement
        Deposit = 2,        // Dépôt à la banque centrale
        Exchange = 3,       // Opération de change
        Adjustment = 4,     // Ajustement
        Transfer = 5        // Transfert entre succursales
    }

    public class CurrencyMovement
    {
        public Guid Id { get; set; }

        [Required]
        public Guid CurrencyReserveId { get; set; }

        [Required]
        public CurrencyMovementType MovementType { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal Amount { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal BalanceBefore { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal BalanceAfter { get; set; }

        [StringLength(100)]
        public string Reference { get; set; } = string.Empty;

        public Guid? ExchangeTransactionId { get; set; }

        [Required]
        [StringLength(500)]
        public string Description { get; set; } = string.Empty;

        [StringLength(500)]
        public string Notes { get; set; } = string.Empty;

        [Required]
        public DateTime MovementDate { get; set; }

        [Required]
        [StringLength(100)]
        public string ProcessedBy { get; set; } = string.Empty;

        [Required]
        [StringLength(100)]
        public string ProcessedByName { get; set; } = string.Empty;

        [Required]
        public DateTime CreatedAt { get; set; }

        // Navigation Properties
        public virtual CurrencyReserve CurrencyReserve { get; set; } = null!;
        public virtual ExchangeTransaction? ExchangeTransaction { get; set; }
    }
}